### PR TITLE
feat: DexScreener (main)

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

Backporting to main: https://github.com/osmosis-labs/osmosis/pull/8386

Note that this is primarily work by @cryptomatictrader which I backport. See original PR ⏫

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -331,21 +339,51 @@ func NewOsmosisApp(
...
+
+	// Initialize the config object for the indexer
+	indexerConfig := indexer.NewConfigFromOptions(appOpts)
+
+	// initialize indexer if enabled
+	if indexerConfig.IsEnabled {
+		indexerIngester := indexerConfig.Initialize()
+
+		// TODO: handle graceful shutdown
+		pubSubCtx := context.Background()
```
Comment (p0mvn) [inline app/app.go:359]: Note: let's keep this in a task to come back to

File: ingest/indexer/indexer_config.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+// Config defines the config for the indexer.
+type Config struct {
+	IsEnabled                bool   `mapstructure:"enabled"`
+	GCPProjectId             string `mapstructure:"gcp-project-id"`
+	BlockTopicId             string `mapstructure:"block-topic-id"`
+	TransactionTopicId       string `mapstructure:"transaction-topic-id"`
+	PoolTopicId              string `mapstructure:"pool-topic-id"`
+	TokenSupplyTopicId       string `mapstructure:"token-supply-topic-id"`
+	TokenSupplyOffsetTopicId string `mapstructure:"token-supply-offset-topic-id"`
+}
```
Comment (p0mvn) [inline ingest/indexer/indexer_config.go:20]: We also need to add logic for generating the default config when we do `osmosisd init`.

It doesn't seem to me that that logic is present. I suspect that's the case for SQS as well.

Let's discuss doing it in this PR vs follow-up

File: ingest/indexer/service/indexer_streaming_service.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+			// Publish the token supply
+			err = s.client.PublishTokenSupply(sdkCtx, indexerdomain.TokenSupply{
+				Denom:  coin.Denom,
+				Supply: coin.Amount,
+			})
+
+			// Skip any error silently but log it.
+			if err != nil {
+				// TODO: alert
+				sdkCtx.Logger().Error("failed to publish token supply", "error", err)
```
Comment (p0mvn) [inline ingest/indexer/service/indexer_streaming_service.go:86]: Create task to add this alert

File: ingest/indexer/service/indexer_streaming_service.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+// ListenCommit updates the steaming service with the latest Commit messages and state changes
+func (s *indexerStreamingService) ListenCommit(ctx context.Context, res abci.ResponseCommit, changeSet []*storetypes.StoreKVPair) error {
+	// If did not ingest initial data yet, ingest it now
+	if !s.coldStartManager.HasIngestedInitialData() {
+		sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+		var err error
+
+		// Ingest the initial data
+		s.keepers.BankKeeper.IterateTotalSupply(sdkCtx, func(coin sdk.Coin) bool {
```
Comment (p0mvn) [inline ingest/indexer/service/indexer_streaming_service.go:71]: Create task to encapsulate and test this

File: ingest/indexer/indexer_config.go (hunk 2 above)
Comment (cryptomatictrader) [reply ingest/indexer/indexer_config.go:20]: Yes I thought about that. Just not sure what default we should have - maybe making them all pointing to dev settings?

File: ingest/indexer/service/indexer_streaming_service.go (hunk 3 above)
Comment (cryptomatictrader) [reply ingest/indexer/service/indexer_streaming_service.go:107]: Do you mean you already created an issue in Linear or you'd like me to create one (part of observability milestone?)

File: ingest/indexer/service/indexer_streaming_service.go (hunk 4 above)
Comment (cryptomatictrader) [reply ingest/indexer/service/indexer_streaming_service.go:92]: Do you mean you already created an issue in Linear or you'd like me to create one?

File: ingest/sqs/domain/mocks/sqs_ingester_mock.go [hunk 5]
```diff
@@ -43,3 +44,7 @@ func (s *SQSIngesterMock) ProcessChangedBlockData(ctx types.Context, changedPool
 	s.LastChangedPoolsObserved = changedPools
 	return s.ChangedBlockDataError
 }
+
+func (SQSIngesterMock) PublishBlock(ctx types.Context, block indexerdomain.Block) error {
+	return nil
+}
```
Comment (p0mvn) [inline ingest/sqs/domain/mocks/sqs_ingester_mock.go:50]: Let's remove this. It's defined on the ingester now AFAIU

File: ingest/indexer/service/indexer_streaming_service.go (hunk 4 above)
Comment (p0mvn) [reply ingest/indexer/service/indexer_streaming_service.go:92]: This was a note to self to not forget. But I haven't gotten to it yet. Would appreciate if you could help with issues if you have time 🙏

File: ingest/indexer/service/indexer_streaming_service.go (hunk 3 above)
Comment (p0mvn) [reply ingest/indexer/service/indexer_streaming_service.go:107]: https://github.com/osmosis-labs/osmosis/pull/8411#discussion_r1649071670

File: ingest/indexer/indexer_config.go (hunk 2 above)
Comment (p0mvn) [reply ingest/indexer/indexer_config.go:20]: I think we can default to disabled and empty strings/zeroes

Let's make sure we track this via issues 🙏

File: app/app.go (hunk 1 above)
Comment (p0mvn) [reply app/app.go:358]: https://github.com/osmosis-labs/osmosis/pull/8411#discussion_r1649071670

File: ingest/sqs/domain/mocks/sqs_ingester_mock.go (hunk 5 above)
Comment (cryptomatictrader) [reply ingest/sqs/domain/mocks/sqs_ingester_mock.go]: Done

File: app/app.go (hunk 1 above)
Comment (cryptomatictrader) [reply app/app.go:358]: [Linear issue created and assigned](https://linear.app/osmosis/issue/IND-50/handle-graceful-shutdown-for-pubsubctx)

File: ingest/indexer/indexer_config.go (hunk 2 above)
Comment (cryptomatictrader) [reply ingest/indexer/indexer_config.go:20]: [Issue created](https://linear.app/osmosis/issue/IND-51/added-defaultconfig-for-sqs-and-indexer-when-osmosisd-init)

File: ingest/indexer/service/indexer_streaming_service.go (hunk 4 above)
Comment (cryptomatictrader) [reply ingest/indexer/service/indexer_streaming_service.go:92]: [Issue created](https://linear.app/osmosis/issue/IND-52/encapsulate-and-test-iteratetotalsupply)

File: ingest/indexer/service/indexer_streaming_service.go (hunk 3 above)
Comment (cryptomatictrader) [reply ingest/indexer/service/indexer_streaming_service.go:107]: [Issue created](https://linear.app/osmosis/issue/IND-53/added-monitoring-point-failed-to-publish-token-supply)

**Review by cryptomatictrader — APPROVED** [review]

**Pull request merged**